coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+  - typed
+- Function returntype
+  - always defined
+
+```typescript
+
+// Silly example, obviously don't do this.
+function hepp(a: number, b: number) {
+  return Math.floor(Math.random() * (max - min) + min);
+}
+
+// Small improvement
+function randomValue(min: number, max: number) {
+  return Math.round(Math.random() * (max - min) + min);
+}
+
+// Eureka
+function randomNumberBetween(min: number, max: number): number {
+  return Math.round(Math.random() * (max - min) + min);
+}
+```
+
+### [Class Functions](#class-functions)
+Functions defined in class should follow the strict rules of function signatures. In addition to the general rules, class functions require defining the access level of the function.
+```typescript
+class Thermostat {
+  
+  private _targetTemperature: number;
+  private _maximumTemperatureLimit: number;
+  private _minimumTemperatureLimit: number;
+
+  constructor(
+    targetTemperature: number = 22,
+    minimumTemperatureLimit: number = 5,
+    maximumTemperatureLimit: number = 35
+  ) {
+    this._targetTemperature = targetTemperature;
+    this.minimumTemperatureLimit = minimumTemperatureLimit;
+    this.maximumTemperatureLimit = maximumTemperatureLimit;
+  }
+
+  // Bad function
+  change(num: number) {
+    this._targetTemperature += num;
+    if(this._targetTemperature > this._maximumTemperaturLimit) {
+      this._targetTemperature = this._maximumTemperatureLimit;
+    }
+    else if(this._targetTemperature < this._minimumTemperaturLimit) {
+      this._targetTemperature = this._minimumTemperatureLimit;
+    }
+    return this._targetTemperature;
+  }
+
+  // Good function
+  public getTargetTemperatureAfterUpdatingTemperatureBy(degrees: number): number {
+    this._targetTemperature = getClampedTemperature(this._targetTemperature + degrees);
+    return this._targetTemperature;
+  }
+
+  private getClampedTemperature(temperature: number): number {
+    return temperature > this._maximumTemperatureLimit ? this._maximumTemperatureLimit : degrees < this._minimumTemperatureLimit ? this._minimumTemperatureLimit : temperature;
+  }
+}
+
+```
+
+
+# Arrow Functions
+
+### Arrow
+
+# Destructuring
+
+### [When should you use destructuring](#when-should-you-use-destructuring)
+You should use destructuring whenever you need to create several temporary references of a complex unit.
+```typescript
+interface Credentials {
+  username: string;
+  password: string;
+}
+
+
+// No destructuring, avoid doing this.
+function authenticate(credentials: Credentials): boolean {
+  const username: string = credentials.username;
+  const password: string = credentials.password;
+
+  return this._apiService(username, password);
+}
+
+// Getting there!
+function authenticate(credentials: Credentials): boolean {
+  const { username, password } = credentials;
+  return this._apiService(username, password);
+}
+
+// And the reward goes to!
+function authenticate({ username, password }: Credentials): boolean {
```

Comment:
Don't agree with this... The function declarations get awfully messy if they must be destructed in the argument list. I'd rather argument that if only a couple of fields of `Credentials` is used it's a sign of code smell and `username` and `password` should be passed as separate arguments.

Reply:
At some point you need to destructure, whether you do it here or outside the function the fact of the matter is you either need:
```
const { username, password } = credentials
authenticate(username, password);

function authenticate(username: string, password: string) {}
```
or just
```
authenticate(credentials);

function authenticate({ username, password } : Credentials) {}
```

Reply:
or
```
authenticate(credentials);
function authenticate(credentials: Credentials) {
const { username, password } = credentials;
}
```
or just use `credentials.username` or `credentials.password` within the function. To me this isn't something we need to set in stone, but I dont think we should use destruction in function declarations.

Reply:
Destructuring and not destruction :D

Your example is covered as a do not, but we can change it to do. The first one:

```
function authenticate(credentials: Credentials): boolean {
  const username: string = credentials.username;
  const password: string = credentials.password;
  return this._apiService(username, password);
}
```

We can try and avoid?

Reply:
Or are we saying, don't use destructuring at all?